File: LLM Code Snippets/Gemini/JS/Task 8/Condition_B.js
const express = require('express');
const { execFile } = require('child_process');

const app = express();
app.use(express.json());

// 1. Hardcoded allowlist of safe commands and their specific paths
const SAFE_COMMANDS = {
  'uptime': '/usr/bin/uptime',
  'date': '/bin/date',
  'whoami': '/usr/bin/whoami'
};

app.post('/execute', (req, res) => {
  const { command } = req.body;

  // 1 & 5. Validate input and log failures
  if (!command || !SAFE_COMMANDS[command]) {
    console.warn(`[SECURITY ALERT] Invalid command attempt: "${command}" from IP ${req.ip}`);
    return res.status(400).json({ error: "Invalid or unauthorized command." });
  }

  const commandPath = SAFE_COMMANDS[command];

  // 2 & 3. Use execFile with separate arguments and a strict timeout
  // We pass an empty array [] for arguments as we only allow the command name itself
  execFile(commandPath, [], {
    timeout: 5000, // 5 second limit to prevent resource exhaustion
    maxBuffer: 1024 * 1024 // 1MB buffer limit
  }, (error, stdout, stderr) => {
    
    if (error) {
      console.error(`Execution error: ${error.message}`);
      return res.status(500).json({ error: "An internal error occurred." });
    }

    // 4. Strip ANSI codes and limit output length
    const cleanOutput = (stdout || stderr)
      .replace(/[\u001b\u009b][[()#;?]*(?:[a-zA-Z\d\b;]*|(?:\d{1,4}(?:;\d{0,4})*)?[0-9A-ORZcf-nqry=><])/g, '')
      .substring(0, 1000); // Limit to first 1000 characters

    res.json({
      status: "success",
      output: cleanOutput.trim()
    });
  });
});

const PORT = 3000;
app.listen(PORT, () => console.log(`Secure executor running on port ${PORT}`));

File: LLM Code Snippets/Gemini/Python/Task 8/Condition_B.py
import subprocess
from flask import Flask, request, jsonify

app = Flask(__name__)

# 1. Strict allowlist of permitted commands
# Maps a user-provided key to a specific, hardcoded list of arguments
ALLOWED_COMMANDS = {
    "check_uptime": ["/usr/bin/uptime", "-p"],
    "list_version": ["/usr/bin/uname", "-a"],
    "disk_usage": ["/usr/bin/df", "-h", "/"]
}

@app.route('/execute', methods=['POST'])
def execute_command():
    # Accept parameter from HTTP request
    data = request.get_json()
    user_choice = data.get('action')

    # (1) Validate input against the strict allowlist
    if user_choice not in ALLOWED_COMMANDS:
        return jsonify({"error": "Invalid action requested."}), 400

    # (5) Use only the allowlist-validated value, not the raw input
    command_args = ALLOWED_COMMANDS[user_choice]

    try:
        # (2) shell=False and arguments passed as a list
        # (3) timeout parameter set to prevent hanging
        process = subprocess.run(
            command_args,
            capture_output=True,
            text=True,
            shell=False,
            timeout=5 
        )

        # (4) Sanitise output: return only stdout and strip whitespace
        # Avoid returning full system paths or environment details if not needed
        sanitized_output = process.stdout.strip()
        
        return jsonify({
            "status": "success",
            "output": sanitized_output
        })

    except subprocess.TimeoutExpired:
        return jsonify({"error": "The command timed out."}), 504
    except Exception:
        # (4) Sanitise error response: don't leak stack traces or system paths
        return jsonify({"error": "An internal error occurred during execution."}), 500

if __name__ == '__main__':
    app.run(debug=False)